src/spacy/domain.clj

```diff
@@ -92,6 +92,46 @@
         (update ::waiting-queue conj new-session)
         (update ::facts into new-facts))))
 
+(defn is-first-in-queue? [state id]
+  (let [next-up (first (get-in state [::waiting-queue]))]
+    (= (get-in next-up [::session ::id]) id)))
+
+(defn is-valid-slot? [state room time]
+  (and (some #{room} (::rooms state))
+       (some #{time} (::times state))))
+
+(defn slot-taken? [state room time]
+  (some (fn [s] (and (= (::room s) room)
+                     (= (::time s) time))) (::schedule state)))
+
+(defn is-open-slot? [state room time]
+  (and (is-valid-slot? state room time)
+       (not (slot-taken? state room time))))
+
+(defn can-schedule-session?
+  "The slot must be open and the session must be the first in the queue"
+  [state {:keys [id room time]}]
+  (and (is-first-in-queue? state id)
+       (is-open-slot? state room time)))
+
+(defn schedule-session [state {:keys [id room time] :as data}]
+  (if-not (can-schedule-session? state data)
+    state ;; if no session can be scheduled, return state with no modification
```

Comment:
I wanted to ensure that what ever is returned from this function is still a valid state, so if `schedule-session` is called with invalid arguments, this just returns the previous state. Since we will usually will directly persist what comes out of these functions, I wanted to make sure it is a valid state.

For the handler function we will have to make sure that we call `can-schedule-event?` before this function so that we can deny any invalid requests.